components/script/dom/readablestream.rs

```diff
@@ -0,0 +1,517 @@
+                self.js_reader.handle(),
+            ));
+            Promise::new_with_js_promise(promise_obj.handle(), cx)
+        }
+    }
+
+    /// Releases the lock on the reader,
+    /// must be done after `start_reading`.
+    #[allow(unsafe_code)]
+    pub fn stop_reading(&self) {
+        if !self.has_reader.get() {
+            panic!("ReadableStream::stop_reading called on a readerless stream.");
+        }
+
+        self.has_reader.set(false);
+
+        let global = self.global();
+        let _ar = enter_realm(&*global);
+        let cx = global.get_cx();
+
+        unsafe {
+            ReadableStreamReaderReleaseLock(*cx, self.js_reader.handle());
+            // Note: is this the way to nullify the Heap?
+            self.js_reader.set(ptr::null_mut());
+        }
+    }
+
+    #[allow(unsafe_code)]
+    pub fn is_locked(&self) -> bool {
+        // If we natively took a reader, we're locked.
+        if self.has_reader.get() {
+            return true;
+        }
+
+        // Otherwise, still double-check that script didn't lock the stream.
+        let cx = self.global().get_cx();
+        let mut locked_or_disturbed = false;
+
+        unsafe {
+            rooted!(in(*cx) let stream = self.js_stream.get());
+            ReadableStreamIsLocked(*cx, stream.handle().into_handle(), &mut locked_or_disturbed);
+        }
+
+        locked_or_disturbed
+    }
+
+    #[allow(unsafe_code)]
+    pub fn is_disturbed(&self) -> bool {
+        // Check that script didn't disturb the stream.
+        let cx = self.global().get_cx();
+        let mut locked_or_disturbed = false;
+
+        unsafe {
+            rooted!(in(*cx) let stream = self.js_stream.get());
+            ReadableStreamIsDisturbed(*cx, stream.handle().into_handle(), &mut locked_or_disturbed);
+        }
+
+        locked_or_disturbed
+    }
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn request_data(
+    source: *const c_void,
+    cx: *mut JSContext,
+    stream: HandleObject,
+    desired_size: usize,
+) {
+    let source = &*(source as *const ExternalUnderlyingSourceController);
+    source.pull(SafeJSContext::from_ptr(cx), stream, desired_size);
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn write_into_read_request_buffer(
+    source: *const c_void,
+    cx: *mut JSContext,
+    stream: HandleObject,
+    buffer: *mut c_void,
+    length: usize,
+    bytes_written: *mut usize,
+) {
+    let source = &*(source as *const ExternalUnderlyingSourceController);
+    source.write_into_buffer(
+        SafeJSContext::from_ptr(cx),
+        stream,
+        buffer,
+        length,
+        bytes_written,
+    );
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn cancel(
+    _source: *const c_void,
+    _cx: *mut JSContext,
+    _stream: HandleObject,
+    _reason: HandleValue,
+) -> *mut JSVal {
+    ptr::null_mut()
+}
+
+#[allow(unsafe_code)]
+unsafe extern "C" fn close(_source: *const c_void, _cx: *mut JSContext, _stream: HandleObject) {}
```

Comment:
Should we clear the buffer vector here? What about setting the `closed` field?

Reply:
I don't think we need to clear the vector there, I think it will drop alongside the entire struct later(there is also a `finalize` callback). 

The `closed` field is to track when native code has closed the source, but the stream itself could not be closed because it wasn't readable at that time(I've tried and you get an error if you do close it when it's not readable). 

So this is rather about if JS were to close the stream, and we could handle that in some way, for example by stopping a native push source.